File: src/main/java/org/retrolauncher/app/platforms/infrastructure/database/jackson/repositories/MemoryPlatformRepository.java
package org.retrolauncher.app.platforms.infrastructure.database.jackson.repositories;

import org.retrolauncher.app.platforms.domain.entities.Platform;
import org.retrolauncher.app.platforms.domain.repositories.PlatformRepository;
import org.retrolauncher.app.platforms.infrastructure.database.jackson.mappers.JacksonPlatformMapper;
import org.retrolauncher.app.platforms.infrastructure.database.jackson.models.PlatformModel;

import java.util.*;

public class MemoryPlatformRepository implements PlatformRepository {
    private final Map<String, PlatformModel> storedData = new HashMap<>();

    @Override
    public void save(Platform entity) {
        this.storedData.put(entity.getId().toString(), JacksonPlatformMapper.fromDomain(entity));
    }

    @Override
    public void clear() {
        this.storedData.clear();
    }

    @Override
    public Optional<Platform> findById(UUID id) {
        PlatformModel model = this.storedData.get(id.toString());
        if (model == null)
            return Optional.empty();
        return Optional.of(JacksonPlatformMapper.toDomain(model));
    }

    @Override
    public List<Platform> listAll() {
        return this.storedData
                .values()
                .stream()
                .map(JacksonPlatformMapper::toDomain).toList();
    }

    @Override
    public boolean existsByCore(String corePath) {
        return this.storedData
                .values()
                .stream()
                .anyMatch((platformModel -> platformModel.getCorePath().equals(corePath)));
    }
}

File: backend/tests/app/platform/doubles/memory_platform_repository.go
package platform_doubles_test

import platform "retrolauncher/backend/src/app/platform/domain"

type MemoryPlatformRepository struct {
	platforms map[string]*platform.Platform
}

func (r *MemoryPlatformRepository) Save(entity *platform.Platform) error {
	if r.platforms == nil {
		r.platforms = make(map[string]*platform.Platform)
	}
	r.platforms[entity.GetID().String()] = entity
	return nil
}

func (r *MemoryPlatformRepository) List() ([]*platform.Platform, error) {
	if r.platforms == nil {
		return make([]*platform.Platform, 0), nil
	}

	result := make([]*platform.Platform, 0, len(r.platforms))
	for _, p := range r.platforms {
		result = append(result, p)
	}
	return result, nil
}

func (r *MemoryPlatformRepository) Size() int {
	if r.platforms == nil {
		return 0
	}
	return len(r.platforms)
}
